megablocks/layers/mlp.py

```diff
@@ -153,17 +154,43 @@ class MemoryOptimizedMLP(torch.autograd.Function):
 
     @staticmethod
     @torch.cuda.amp.custom_fwd
-    def forward(ctx, x, w1, w2, topo):
+    def forward(ctx, x, w1, w2, topo, num_bits):
         # x: [m, k], w1: [n, k], w2: [n, k]
         if (not x.is_contiguous() or not w1.is_contiguous() or
             not w2.is_contiguous()):
             raise ValueError("Expected contiguous 'x', 'w1' and 'w2'.")
+        if num_bits not in (-1, 4, 8):
```

Comment:
Can we do this error checking more centrally? You could put it in a helper and then validate in the MoE/dMoE layers constructors?